k2/csrc/fsa_utils.cu

```diff
@@ -99,75 +99,237 @@ static void SplitStringToVector(const std::string &in, const char *delim,
+
+  return fsa;
+}
+
+/* Create an acceptor from a stream, assuming the acceptor is in the OpenFST
+   format:
+
+   src_state1 dest_state1 label1 score1
+   src_state2 dest_state2 label2 score2
+   ... ...
+   final_state final_score
+
+   We will negate the cost/score when we read them in. Also note, OpenFST may
+   omit the cost/score if it is 0.0.
+
+   If there are multiple final states, we will create a super final state, and
+   create arcs from the old final states to the new super final state, with the
+   (negated) old final state cost/score as its cost/score, and -1 as its label.
+
+   @param [in]  is    The input stream that contains the acceptor.
+
+   @return It returns an Fsa on CPU.
+*/
+static Fsa OpenFSTAcceptorFromStream(std::istringstream &is) {
+  std::vector<Arc> arcs;
+  std::vector<std::vector<Arc>> state_to_arcs;            // indexed by states
+  std::vector<std::string> splits;
+  std::string line;
 
   int32_t max_state = -1;
+  int32_t num_arcs = 0;
   std::vector<int32_t> original_final_states;
   std::vector<float> original_final_weights;
-  do {
+  while (std::getline(is, line)) {
     SplitStringToVector(line, kDelim,
                         &splits);  // splits is cleared in the function
     if (splits.empty()) continue;  // this is an empty line
 
     auto num_fields = splits.size();
-
-    if (num_fields == 4u) {
+    if (num_fields == 3u || num_fields == 4u) {
+      //   0            1          2
+      // src_state  dest_state   label
+      //
+      // or
+      //
       //   0            1          2      3
-      // src_state  dest_state  symbol  score
+      // src_state  dest_state   label  score
       int32_t src_state = StringToInt(splits[0]);
       int32_t dest_state = StringToInt(splits[1]);
       int32_t symbol = StringToInt(splits[2]);
-      float score = scale * StringToFloat(splits[3]);
-      arcs.emplace_back(src_state, dest_state, symbol, score);
+      float score = 0.0f;
+      if (num_fields == 4u)
+        score = -1.0f * StringToFloat(splits[3]);
+
+      // Add the arc to "state_to_arcs".
+      ++num_arcs;
       max_state = std::max(max_state, std::max(src_state, dest_state));
-    } else if (num_fields == 2u) {
+      if (static_cast<int32_t>(state_to_arcs.size()) <= src_state)
+        state_to_arcs.resize(src_state + 1);
+      state_to_arcs[src_state].emplace_back(src_state, dest_state, symbol,
+                                            score);
+    } else if (num_fields == 1u || num_fields == 2u) {
       //   0            1
       // final_state  score
-      // In this case, openfst is true. There could be multiple final states, so
-      // we first have to collect all the final states, and then work out the
-      // super final state.
-      K2_CHECK(openfst) << "Invalid line: " << line
-                        << "\nFinal state with weight detected in K2 format";
+      float score = 0.0f;
+      if (num_fields == 2u)
+        score = StringToFloat(splits[1]);
```

Comment:
the scale -1 is missing.

Reply:
The variable is called `original_final_weights` so I didn't negate it here. I did it later: https://github.com/k2-fsa/k2/blob/04872431289b5b3e9c154ef389d094d08a1a6e9c/k2/csrc/fsa_utils.cu#L316

But I can move it up if that's easier to read.

Reply:
Please add a comment here or move it up. That's easier to read.